syntax = "proto3";
package cle.proto;

enum CardType {
    CARD_TYPE_CREATURE = 0;
    CARD_TYPE_INSTANT = 1;
    CARD_TYPE_SORCERY = 2;
    CARD_TYPE_ENCHANTMENT = 3;
    CARD_TYPE_ARTIFACT = 4;
    CARD_TYPE_PLANESWALKER = 5;
    CARD_TYPE_LAND = 6;
}

enum ManaColor {
    MANA_COLOR_COLORLESS = 0;
    MANA_COLOR_WHITE = 1;
    MANA_COLOR_BLUE = 2;
    MANA_COLOR_BLACK = 3;
    MANA_COLOR_RED = 4;
    MANA_COLOR_GREEN = 5;
}

message HybridCost {
    ManaColor primary = 1;
    ManaColor secondary = 2;
}

message ManaCost {
    int32 colorless = 1;
    int32 white = 2;
    int32 blue = 3;
    int32 black = 4;
    int32 red = 5;
    int32 green = 6;
    int32 x_count = 7;
    repeated HybridCost hybrid_costs = 8;
}

message CreatureStats {
    int32 power = 1;
    int32 toughness = 2;
}

message ActivatedAbilityData {
    string cost_text = 1;
    string effect_text = 2;
    bool sorcery_speed_only = 3;
}

message StaticAbilityData {
    string description = 1;
}

message ModalChoiceData {
    string text = 1;
}

message ModalAbilityData {
    int32 min_choices = 1;
    int32 max_choices = 2;
    repeated ModalChoiceData modes = 3;
}

message CardData {
    string name = 1;
    CardType type = 2;
    ManaCost mana_cost = 3;
    uint32 colors = 4;
    string oracle_text = 5;
    string flavor_text = 6;
    repeated string subtypes = 7;
    repeated string keywords = 8;
    optional CreatureStats creature_stats = 9;
    uint64 instance_id = 10;
    repeated int32 trigger_types = 11;
    repeated ActivatedAbilityData activated_abilities = 12;
    repeated StaticAbilityData static_abilities = 13;
    optional ModalAbilityData modal = 14;
    string lua_source_path = 15;
}
